lib/Adapter/Tolerant/TolerantIndexBuilder.php

```diff
@@ -0,0 +1,155 @@
+            return;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $this->indexNode($info, $childNode);
+        }
+    }
+
+    private function indexClassDeclaration(SplFileInfo $info, ClassDeclaration $node): void
+    {
+        $record = $this->index->get(ClassRecord::fromName($node->getNamespacedName()->getFullyQualifiedNameText()));
+        assert($record instanceof ClassRecord);
+        $record->withLastModified($info->getCTime());
+        $record->withStart(ByteOffset::fromInt($node->getStart()));
+        $record->withType('class');
+        $record->withFilePath($info->getPathname());
+
+        // de-reference this class
+        foreach ($record->implementedClasses() as $implementedClass) {
+            $implementedRecord = $this->index->get(ClassRecord::fromName($implementedClass));
+
+            if (false === $implementedRecord->removeImplementation($record->fqn())) {
+                continue;
+            }
+
+            $this->index->write($implementedRecord);
+        }
+
+        $record->clearImplements();
+
+        $this->indexClassInterfaces($record, $node);
+        $this->indexBaseClass($record, $node);
+
+        $this->index->write($record);
+    }
+
+    private function indexClassInterfaces(ClassRecord $classRecord, ClassDeclaration $node): void
+    {
+        // @phpstan-ignore-next-line
+        if (null === $interfaceClause = $node->classInterfaceClause) {
+            return;
+        }
+
+        if (null == $interfaceList = $interfaceClause->interfaceNameList) {
+            return;
+        }
+
+        foreach ($interfaceList->children as $interfaceName) {
+            if (!$interfaceName instanceof QualifiedName) {
+                continue;
+            }
+
+            $classRecord->addImplements(FullyQualifiedName::fromString($interfaceName->getNamespacedName()->getFullyQualifiedNameText()));
+
+            $interfaceRecord = $this->index->get(ClassRecord::fromName($interfaceName));
+            assert($interfaceRecord instanceof ClassRecord);
+            $interfaceRecord->addImplementation($classRecord->fqn());
+
+            $this->index->write($interfaceRecord);
+        }
+    }
+
+    private function indexBaseClass(ClassRecord $record, ClassDeclaration $node): void
+    {
+        // @phpstan-ignore-next-line
+        if (null === $baseClause = $node->classBaseClause) {
+            return;
+        }
+
+        // @phpstan-ignore-next-line
+        if (null === $baseClass = $baseClause->baseClass) {
+            return;
+        }
+
+        $name = $baseClass->getNamespacedName()->getFullyQualifiedNameText();
+        $record->addImplements(FullyQualifiedName::fromString($name));
+        $baseClassRecord = $this->index->get(ClassRecord::fromName($name));
+        assert($baseClassRecord instanceof ClassRecord);
+        $baseClassRecord->addImplementation($record->fqn());
+        $this->index->write($baseClassRecord);
```

Comment:
Close duplicate of what is done in the `foreach` above for interfaces.
Maybe refactor it or consider the rule of three and let it pass if you don't think it will change often or be used somewhere else again.

Reply:
Maybe in the future